starfish/pipeline/features/pixels/combine_adjacent_features.py

```diff
@@ -0,0 +1,117 @@
+from typing import List, Tuple, Union
+
+import numpy as np
+import pandas as pd
+from skimage.measure import regionprops, label
+
+from starfish.munge import dataframe_to_multiindex
+from starfish.constants import Indices
+from starfish.intensity_table import IntensityTable
+
+Number = Union[int, float]
+
+
+def combine_adjacent_features(
+        intensities: IntensityTable, area_threshold: Number, connectivity: int=2,
+) -> Tuple["IntensityTable", List, np.ndarray, np.ndarray]:
+    """
+
+    Parameters
+    ----------
+    intensities : IntensityTable
+        pixel-valued IntensityTable
+    area_threshold : Number
+        discard spots whose total pixel area is less than this value
+    connectivity : int
+        the connectivity allowed when constructing connected components. Intuitively it can be
+        thought of as the number of square steps allowed between pixels that are considered
+        connected. 1 enforces adjacent pixels. 2 (default) allows diagonal pixels in the same
+        plane. 3 allows diagonal pixels in 3d.
+
+    Returns
+    -------
+    IntensityTable :
+        spot intensities
+    List[Object] :
+        region properties for each spot (see skimage.measure.regionprops)
+    np.ndarray[int] :
+        label image wherein each connected component (spot) is coded with a different
+        integer
+    np.ndarray[int] :
+        decoded image wherein each target is coded as a different integer. Intended for visualization
+
+    """
+    # None needs to map to zero, non-none needs to map to something else.
+    int_to_gene = dict(
+        zip(range(1, np.iinfo(np.int).max),
+            set(intensities.gene_name.values) - {'None'}))
+    int_to_gene[0] = 'None'
+    gene_to_int = {v: k for (k, v) in int_to_gene.items()}
+
+    # map genes to ints
+    gene_list = [gene_to_int[g] for g in intensities.coords[intensities.Constants.GENE.value].values]
+    gene_array = np.array(gene_list)
+
+    # reverses the linearization that was used to construct the IntensityTable from the ImageStack
+    decoded_image: np.ndarray = gene_array.reshape(intensities.attrs['image_shape'])
+
+    # label each pixel according to its component
+    label_image: np.ndarray = label(decoded_image, connectivity=connectivity)
+
+    # calculate the mean value grouped by (ch, round) across all pixels of a connected component
+    pixel_labels = label_image.reshape(-1)
+    intensities['spot_id'] = (IntensityTable.Constants.FEATURES.value, pixel_labels)
+    mean_pixel_traces = intensities.groupby('spot_id').mean('features')
+
+    # "Zero" spot id is background in the label_image. Remove it.
+    mean_pixel_traces = mean_pixel_traces.loc[mean_pixel_traces['spot_id'] > 0]
```

Comment:
is there a simpler way to do this? perhaps up above

Reply:
Added TODO.